README.md
 
Incorporated Research Institutions for Seismology (IRIS)
Data Management Center (DMC)
Data Products Team
IRIS Earth Model Collaboration (EMC) - ParaView support bundle

COMMENTS/QUESTIONS:

    Please contact manoch@iris.washington.edu

Python 3
2021-10-04
------------------------------------------------------------------------------------------------------------------------

 DESCRIPTION:

 This package provides a set of Python 3 programmable filters/sources to allow ParaView open-source, multi-platform data 
 analysis and visualization application (on macOS®, Linux®, and Windows® platforms) to display EMC netCDF/GeoCSV models 
 along with other auxiliary Earth data.  After installing this package (see INSTALL.txt) the plugins/macros in the 
 package will allow you to:

    - Draw Boundaries of:
        + very low-resolution, low-resolution or intermediate resolution coastlines
        + national, US states and Canadian provinces 
        + present-day plates (divergent, transform or convergent margins)
        + your own GeoCSV boundary files

    - Show Volcano Locations:
        + plot location of volcanoes using WOVOdat location data from ds.iris.edu
        + plot location of volcanoes using your GeoCSV volcano location or other data-point files

    - Show Earthquake Locations:
        + plot earthquake locations based on the FDSN event services
        + plot earthquake locations based on your local GeoCSV earthquake location file
 
    NOTE: For information on GeoCSV files, visit: http://geows.ds.iris.edu/documents/GeoCSV.pdf
    NOTE: We also offer Python scripts for converting model files from/to netCDF and GeoCSV formats.
          For more information, please visit: https://github.com/iris-edu/emc-tools
    NOTE: Unless data files are available locally, an Internet connection is required for data download
  
    - Show USGS Slab 1.0 - Display USGS Slab 1.0 models 
    - Draw grids - Draw latitude/longitude grid lines
    - Elevation data - Plot surface elevation changes using elevation data files 
    - Read Earth Models - Read and plot 2D and 3D Earth model files in netCDF and GeoCSV file format. 
                          The available EMC model files are at https://ds.iris.edu/files/products/emc/emc-files/
                          The model descriptions are available at http://ds.iris.edu/ds/products/emc-earthmodels/
      NOTE: You do not need to download the model files. Just insert the name of the file into the "FileName" 
            property box
    - Convert latitude, longitude and, depth to X, Y, and Z (Or Vice Versa) 
    - Display the coordinates (latitude, longitude, and depth) of an object in the Pipeline Browser.


 USAGE:
    You can work with plugins directly, or you can work with macros. Macros are wrappers for plugins that include 
    additional steps needed to tune the display. 

    PLUGINS:
    - From the "Sources --> IRIS EMC" menu, select the plugin of interest
    - The Area drop-down allows you to limit the plot to a particular geographic area, or you can define your region 
      by specifying the lat/lon limits

      NOTE: by populating any of the latitude or longitude limit boxes, you are modifying the corresponding limit of 
      the selected area

    - Click on Apply button to run the plugin. 
         + The following plugins will provide a display that you can tune by changing line, color, etc.:
              Draw Boundaries, Draw grids, Show USGS Slab 1.0 and Topo elevation data

         + The following plugins are of point type, and you need to instruct ParaView how to display them before you can 
            tune them by changing line, color, etc.:
              Show Earthquake Locations - set Representation to "Point Gaussian," and set Gaussian Radius to a smaller 
                number like 0.01
              Show Volcano Locations - set Representation to "Point Gaussian," Shared Preset to "Triangle," and set 
                Gaussian Radius to a smaller number like 0.005
              
         + The following plugin is of surface type, and you need to instruct ParaView how to display it before you can 
               tune them by changing line, color, etc.:
                   Read 2D Model and Read 3D Model - set Representation "Surface" and select the variable for coloring 
                   using the drop-down that shows "Solid Color"

    - Two additional plugins serve as convenience tools to obtain the coordinate information:
        + Convert Coordinates - a coordinate conversion tool that allows you to convert ParaView's Cartesian coordinates 
            to latitude, longitude, and depth
        + View Coordinates - to display coordinates (latitude, longitude, and depth) of a pipeline object 
            The objects created in the PipelineBrowser via the EMC plugins preserve the coordinate information. You may 
            review the coordinates under a SpreadSheetView using the drop-down menu. For objects in the PipelineBrowser 
            that are created from EMC objects using one of the ParaView tools like "Slice," select the object in the 
            PipelineBrowser, then in paraview menu select "Filters --> IRIS EMC" that will open a SpreadSheetView and 
            display coordinate information of that object.

    MACROS:
    - From ParaView's Macros menu, select "Add new macro..." and browse to the  IRIS_EMC_Paraview/macros directory and 
        add the macros you want.
    - From ParaView's Macros menu or ParaView's toolbar, click on the macro you want to run
    - Like plugins, adjust the display using the property panel

      NOTE: You can create your macros as described here: https://www.paraview.org/Wiki/Python_GUI_Tools
 


CHANGES.txt

Incorporated Research Institutions for Seismology (IRIS)
Data Management Center (DMC)
Data Products Team
IRIS Earth Model Collaboration (EMC) - ParaView support bundle

COMMENTS/QUESTIONS:

    Please contact manoch@iris.washington.edu

---------------------------------------------------------

=========================================================
2021-10-03 r2.0.0 Python 3
=========================================================
- Python 3 release

=========================================================
2019-05-13 r1.5.6.4
=========================================================
- documents/IRIS_DS_SSA_2019.pdf added V.2019.091 Developing and Visualizing Community Seismic Velocity Models Workshop

=========================================================
2019-04-22 r1.5.6.3
=========================================================
- readers/read_2d_models.py   V.2019.112 Variable "File_name" renamed.
- readers/read_3d_models.py   V.2019.112 PEP8 style update.


=========================================================
2019-04-17 r1.5.6.2
=========================================================

- documents/IRIS_EMC_ParaView_Plugins_Exercise_2_Model.pdf V.2019.091 corrected the inverted latitude begin/end
- IrisEMC_Paraview_Param                                   V.2019.107 usgsSlab_URL is directed to IRIS for now to
                                                           avoid issues with .csv files
- IrisEMC_Paraview_Lib.py                                  V.2019.107 corrected a statement spacing in
                                                           read_geocsv_model_2d and read_geocsv_model_3d
                                                           that was preventing correct data export in some cases

=========================================================
2019-03-26 r1.5.6.1
=========================================================
- readers/read_3d_models.py   2019.079 added support for model file names without extension
- readers/read_2d_models.py   2019.079 added support for model file names without extension
- documents/IRIS_EMC_ParaView_Plugins_Guide.pdf V.2019.091 updated
- documents/IRIS_EMC_ParaView_Plugins_Installation_Guide.pdf V.2019.091 added
- documents/IRIS_EMC_ParaView_Plugins_Exercise_1_EQ.pdf V.2019.091 added
- documents/IRIS_EMC_ParaView_Plugins_Exercise_2_Model.pdf V.2019.091 added

=========================================================
2019-03-01 r1.5.6
=========================================================
- IrisEMC_Paraview_Lib.py     V.2019.059 added missing_value support to read_netcdf_model
- readers/read_3d_models.py   V.2019.060 added the exception error to output message
- readers/read_2d_models.py   V.2019.060 added the exception error to output message

=========================================================
2019-01-30 r1.5.5
=========================================================
- readers/read_earthquakes.py V.2019.031 now produces animation frame files for both single frame and accumulative modes
- documents/IRIS_EMC_ParaView_Plugins_Guide.pdf V.2019.031 Updated earthquake animation section to include both
                                                single frame and accumulative modes
- documents/SINGLE_2015-09-16_NEAR_COAST_OF_CENTRAL_CHILE_09-14_09626.mov V.2019.031 added single frame video
- documents/2015-09-16_NEAR_COAST_OF_CENTRAL_CHILE_09-14_09626.mov V.2019.031 recreated to have the same style as the
                                                                    single frame video

=========================================================
2019-01-30 r1.5.4
=========================================================
- documents/IRIS_EMC_ParaView_Plugins_Guide.pdf V.2019.030 Added
- documents/2015-09-16_NEAR_COAST_OF_CENTRAL_CHILE_09-14_09626.mov V.2019.030 Added
- IrisEMC_Paraview_Utils      V.2019.030 datetime_to_float now accepts date with or without time
- readers/read_3d_models.py   V.2019.022 modified to accept depth_begin and depth_end as floats
- readers/read_earthquakes.py V.2019.022 read_3d_models.py support for float depth begin and end
- IrisEMC_Paraview_Param      V.2019.022 added animation directory under earthquakes path and
                                         introduced time_column for use with earthquake" IrisEMC_Paraview_Param.py
- readers/read_earthquakes.py V.2019.022 added the time_end parameter and renames start_time to time_begin so user
                                         can limit event request to a selected time window. Introduced the Frame_Tag
                                         parameter in support of the animation
                                         reader now can output frames for animation
- IrisEMC_Paraview_Utils      V.2019.022 added datetime_to_int, datetime_to_float, remove_files

=========================================================
2019-01-14 r1.5.3
=========================================================
- IrisEMC_Paraview_Lib.py now gets the default volcano data from IRIS EMC file repository that is a curated
                          WOVOdate data
- IrisEMC_Paraview_Param.py support for getting the default volcano data from IRIS EMC file repository
- macros/macro_volcanoes.py now gets the default volcano data from IRIS EMC file repository
- readers/read_volcanoes.py now gets the default volcano data from IRIS EMC file repository


=========================================================
2018-12-21 r1.5.2
=========================================================
- readers/read_earthquakes.py now applies earthquake parameters to the local files too

=========================================================
2018-12-20 r1.5.2
=========================================================
- filters/filter_view_coordinates.py corrected the bug that would prevent it from outputting data values

=========================================================
2018-12-18 r1.5.2
=========================================================

src/IrisEMC_Paraview_Param.py
import os
import IrisEMC_Paraview_Utils as utils

"""
 NAME: IrisEMC_Paraview-Param.py - EMC ParaView Support parameters

       http://ds.iris.edu/ds/products/emc/

 DESCRIPTION: These parameters are the default values for the IRIS EMC Paraview Python scripts

 Copyright (C) 2021  Product Team, IRIS Data Management Center

    This is a free software; you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as
    published by the Free Software Foundation; either version 3 of the
    License, or (at your option) any later version.

    This script is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License (GNU-LGPL) for more details.  The
    GNU-LGPL and further information can be found here:
    http://www.gnu.org/

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

 HISTORY:
    2021-10-03 Manoch: v.2021.276 Python 3 release r2
    2019-04-17 Manoch: v.2019.107 usgsSlab_URL is directed to IRIS for now to avoid issues with .csv files
    2019-01-30 Manoch: v.2019.030 event service call request order changed from magnitude to time-asc
    2019-01-22 Manoch: v.2019.022 added animation directory under earthquakes path and introduced time_column
                       for use with earthquake
    2019-01-14 Manoch: v.2019.014 support for getting the default volcano data from IRIS EMC file repository
    2018-11-12 Manoch: v.2018.316 added Platform check to load .csv files instead of .nc for Windows
    2018-10-17 Manoch: v.2018.290 updates for R1
    2018-09-13 Manoch: v.2018.256 added support for EMC_DEFAULT_GSV_MODEL
    2018-05-09 Manoch: v.2018.129 added EMC_DEFAULT_2DMODEL tp fileDict
    2018-04-27 Manoch: v.2018.117 updateid irisEMC_Files_URL
    2018-04-23 Manoch: v.2018.113 updateid lat and lon limits for the world
                       added very low resolution coastline data file
    2018-03-21 Manoch: v.2018.080 release
"""


# see if SSL is available for HTTPS requests
ssl_available = utils.do_https()
HTTP_PROTOCOL = 'https:'
if not ssl_available:
    HTTP_PROTOCOL = 'http:'


def sort_dict_by_value(dictionary):
    """Splits a Python dictionary to two lists containing keys and values sorted by values.

    Keyword arguments:
    thisDict: a Python dictionary

    Return values:
    keys: list of keys in the dictionary
    values: list of values in the dictionary
    """
    from operator import itemgetter
    keys = []
    values = []
    if len(list(dictionary.keys())) == 1:
        return list(dictionary.keys()), list(dictionary.values())

    for _key, _value in sorted(list(dictionary.items()), key=itemgetter(1)):
        keys.append(_key)
        values.append(_value)
    return keys, values


# URLs
irisEMC_Files_URL = f"{HTTP_PROTOCOL}//ds.iris.edu/files/products/emc/emc-files/"

# usgsEvent_URL = "%s//earthquake.usgs.gov/fdsnws/event/1/query?format=text" % HTTP_PROTOCOL

if not ssl_available:
    usgsSlab_URL = f"{HTTP_PROTOCOL}//ds.iris.edu/files/products/emc/emc-files/"
else:
    # usgsSlab_URL = "%s//earthquake.usgs.gov/static/lfs/data/slab/models/" % HTTP_PROTOCOL

    # For now we serve the slabs from IRIS to make sure we have full support for the CSV version
    # Manoch 2019-04-17
    usgsSlab_URL = f"{HTTP_PROTOCOL}//ds.iris.edu/files/products/emc/emc-files/"

# earthquake catalogue sources
# Note: at this time only GeoCSV format is supported for earthquake files
# Note: A temporary fix to deal with  SSL issue

if not ssl_available:
    earthquakeCatalogDict = {f'{HTTP_PROTOCOL}//service.iris.edu/fdsnws/event/1/query':
                             'IRIS DMC FDSNWS event Web Service'}
    earthquakeQuery = ("format=text&starttime=%s&endtime=%s&minmag=%0.1f&maxmag=%0.1f&"""
                       "orderby=time-asc&mindepth=%0.1f&"""
                       "maxdepth=%0.1f&minlat=%0.2f&maxlat=%0.2f&minlon=%0.2f&maxlon=%0.2f&nodata=404")
else:
    earthquakeCatalogDict = {f'{HTTP_PROTOCOL}//earthquake.usgs.gov/fdsnws/event/1/query':
                             'USGS Earthquake Hazards Program'}
    earthquakeQuery = ("format=text&starttime=%s&endtime=%s&minmag=%0.1f&maxmag=%0.1f&"""
                       "orderby=time-asc&mindepth=%0.1f&"""
                       "maxdepth=%0.1f&minlat=%0.2f&maxlat=%0.2f&minlon=%0.2f&maxlon=%0.2f&nodata=404")

# sort the earthquake dictionary based on the values (organization)
(earthquakeKeys, earthquakeValues) = sort_dict_by_value(earthquakeCatalogDict)

# Paths
topDir = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))

pathDict = {'EMC_MAIN_PATH': topDir,
            'EMC_PLUGINS_PATH': os.path.join(topDir, 'plugins'),
            'EMC_MACROS_PATH': os.path.join(topDir, 'macros'),
            'EMC_SRC_PATH': os.path.join(topDir, 'src'),
            'EMC_DATA_PATH': os.path.join(topDir, 'data'),
            'EMC_SCRATCH_PATH': os.path.join(topDir, 'data', 'scratch'),
            'EMC_SLABS_PATH': os.path.join(topDir, 'data', 'slabs'),
            'EMC_MODELS_PATH': os.path.join(topDir, 'data', 'models'),
            'EMC_BOUNDARIES_PATH': os.path.join(topDir, 'data', 'boundaries'),
            'EMC_VOLCANOES_PATH': os.path.join(topDir, 'data', 'volcanoes'),
            'EMC_EARTHQUAKES_PATH': os.path.join(topDir, 'data', 'earthquakes'),
            'EMC_EQ_ANIMATION_PATH': os.path.join(topDir, 'data', 'earthquakes', 'animation')}

for key in list(pathDict.keys()):
    utils.makePath(pathDict[key])

"""Default files.
   NOTE: to make sure scipy is available under pvpython, we need to set the extension at run time
         '' is used to flag the need for update"""
filesExtDict = {'ssl': '.nc', 'geo': '.csv'}
filesDict = {'EMC_DEFAULT_MODEL': 'wUS-SH-2010_percent', 'EMC_DEFAULT_2DMODEL': 'CAM2016Litho',
             'EMC_DEFAULT_VOLCANO': 'WOVOdat_volcano_locations.csv'}

# Default column names for GeoCSV files. User can redefine these in the GeoCSV header.
columnKeys = {'latitude_column': 'latitude', 'longitude_column': 'longitude', 'depth_column': 'depth',
              'elevation_column': 'elevation', 'magnitude_column': 'magnitude', 'time_column': 'time'}

""" USGS Slab 1.0 files, will populate the drop down menu.
    NOTE: to make sure scipy is available under pvpython, we need to set the extension at run time
          '' is used to flag the need for update"""
grd_ext = ''
usgsSlabExtDict = {'ssl': '.grd', 'geo': '.csv'}
usgsSlabDict = {'alu_slab1.0_clip%s' % grd_ext: 'Alaska-Aleutians',
                'cas_slab1.0_clip%s' % grd_ext: 'Cascadia',
                'izu_slab1.0_clip%s' % grd_ext: 'Izu-Bonin',
                'ker_slab1.0_clip%s' % grd_ext: 'Kermadec-Tonga',
                'kur_slab1.0_clip%s' % grd_ext: 'Kamchatka/Kurils/Japan',
                'mex_slab1.0_clip%s' % grd_ext: 'Central America',
                'phi_slab1.0_clip%s' % grd_ext: 'Philippines',
                'ryu_slab1.0_clip%s' % grd_ext: 'Ryukyu',
                'sam_slab1.0_clip%s' % grd_ext: 'South America',
                'sco_slab1.0_clip%s' % grd_ext: 'Scotia',
                'sol_slab1.0_clip%s' % grd_ext: 'Solomon Islands',
                'van_slab1.0_clip%s' % grd_ext: 'Santa Cruz Islands/Vanuatu/Loyalty Islands'}

# Sort the SLAB dictionary based on the values (regions).
(usgsSlabKeys, usgsSlabValues) = sort_dict_by_value(usgsSlabDict)

# Extent of the slab regions.
usgsSlabRangeDict = {'alu_slab1.0_clip%s' % grd_ext: {'X': (167., 216), 'Y': (50., 65.), 'Z': (-278.8, -6.264)},
                     'cas_slab1.0_clip%s' % grd_ext: {'X': (-128.5, -120.5), 'Y': (39., 52.), 'Z': (100.18, -4.99)},
                     'izu_slab1.0_clip%s' % grd_ext: {'X': (129., 148.), 'Y': (11., 40.), 'Z': (-686.764, -1.002)},
                     'ker_slab1.0_clip%s' % grd_ext: {'X': (174., 188.), 'Y': (-39., -14.), 'Z': (-700.35, -2.81)},
                     'kur_slab1.0_clip%s' % grd_ext: {'X': (129., 164.), 'Y': (32., 56.5), 'Z': (-724.18, -5.85)},
                     'mex_slab1.0_clip%s' % grd_ext: {'X': (254., 279.), 'Y': (7., 21), 'Z': (-294.83, 0.402)},
                     'phi_slab1.0_clip%s' % grd_ext: {'X': (122., 128.), 'Y': (7, 15), 'Z': (-224.68, -6.42)},
                     'ryu_slab1.0_clip%s' % grd_ext: {'X': (122., 139), 'Y': (22., 38), 'Z': (-287.93, 0.78)},
                     'sam_slab1.0_clip%s' % grd_ext: {'X': (278., 300.), 'Y': (-45., 5.), 'Z': (-742.32, 1.83)},
                     'sco_slab1.0_clip%s' % grd_ext: {'X': (328., 337.), 'Y': (-61., -55.), 'Z': (-260.69, -4.88)},
                     'sol_slab1.0_clip%s' % grd_ext: {'X': (145., 165.), 'Y': (-12., -2.), 'Z': (-616.65, -1.81)},
                     'van_slab1.0_clip%s' % grd_ext: {'X': (164., 173.), 'Y': (-23.5, -9.), 'Z': (-331.3, -3.89)}}

# Areas.
areaDict = {'0': 'Africa',
            '1': 'Americas',
            '2': 'Asia',
            '3': 'Europe',
            '4': 'Japan',
            '5': 'North America',
            '6': 'South America',
            '7': 'Taiwan',
            '8': 'USA',
            '9': 'World'}

# Sort the Area dictionary based on the values (regions).
(areaKeys, areaValues) = sort_dict_by_value(areaDict)

# Extent of the areas.
areaRangeDict = {'0': {'lat': (-35., 40.), 'lon': (-20, 55)},
                 '1': {'lat': (-60., 80.), 'lon': (-170., -30.)},
                 '2': {'lat': (0., 80.), 'lon': (55., 179.9)},
                 '3': {'lat': (35., 75.), 'lon': (-20., 50.)},
                 '4': {'lat': (24., 46.), 'lon': (120., 150.)},
                 '5': {'lat': (10., 80.), 'lon': (-170., -50.)},
                 '6': {'lat': (-60., 20.), 'lon': (-85., -30)},
                 '7': {'lat': (21.5, 25.5), 'lon': (120.0, 122.5)},
                 '8': {'lat': (25., 57.), 'lon': (-125., -66.5)},
                 '9': {'lat': (-90., 90.), 'lon': (-180., 180.)}}

# Boundary files that will appear in the drop down menu.
boundariesDict = {'coastline_data_int.csv': 'Coastline data: intermediate resolution',
                  'coastline_data_low.csv': 'Coastline data: low resolution',
                  'coastline_data_very_low.csv': 'Coastline data: very low resolution',
                  'national_boundaries_data.csv': 'National boundaries + US state + Canadian province boundaries',
                  'plate_boundaries_divergent.csv': 'Present-day plate boundaries: divergent margins',
                  'plate_boundaries_transform.csv': 'Present-day plate boundaries: transform margins',
                  'plate_boundaries_convergent.csv': 'Present-day plate boundaries: convergent margins'}
boundariesColor= {'coastline_data_int.csv': '1,1,1',
                  'coastline_data_low.csv': '1,1,1',
                  'coastline_data_very_low.csv': '1,1,1',
                  'national_boundaries_data.csv': '1,1,1',
                  'plate_boundaries_divergent.csv': '0.0, 0.3333333333333333, 1.0',
                  'plate_boundaries_transform.csv': '1.0, 0.0, 0.0',
                  'plate_boundaries_convergent.csv': '1.0, 1.0, 0.0'}

# Sorted list of the boundary files based on values.
(boundaryKeys, boundaryValues) = sort_dict_by_value(boundariesDict)

"""Boundary files that will appear in the drop down menu
NOTE: to make sure scipy is available under pvpython, we need to set the extension at run time
      '' is used to flag the need for update"""
grd_ext = ''
topoExtDict = {'ssl': '.nc', 'geo': '.csv'}
topoDict = {'GMTED2010_15n240_1000deg_dmc%s' % grd_ext: 'GMTED2010 elevation data: 1.000 degrees resolution',
            'GMTED2010_15n120_0500deg_dmc%s' % grd_ext: 'GMTED2010 elevation data: 0.500 degrees resolution',
            'GMTED2010_15n060_0250deg_dmc%s' % grd_ext: 'GMTED2010 elevation data: 0.250 degrees resolution'}

# sorted list of the topo files based on values
(topoKeys, topoValues) = sort_dict_by_value(topoDict)
